libs/datamodel/core/src/transform/ast_to_dml/db/walkers/relation.rs

```diff
@@ -1,15 +1,136 @@
+use super::{ModelWalker, RelationFieldWalker, ScalarFieldWalker};
+use crate::{
+    ast,
+    transform::ast_to_dml::db::{relations::*, ParserDatabase},
+};
 use datamodel_connector::ConnectorCapability;
 use dml::relation_info::ReferentialAction;
 
-use crate::{
-    ast::{self, FieldArity},
-    transform::ast_to_dml::db::{
-        relations::{Relation, RelationType},
-        ParserDatabase,
-    },
-};
+#[derive(Copy, Clone)]
+pub(crate) struct RelationWalker<'ast, 'db> {
+    pub(super) relation_id: usize,
+    pub(super) db: &'db ParserDatabase<'ast>,
+}
 
-use super::{ModelWalker, RelationFieldWalker, ScalarFieldWalker};
+impl<'ast, 'db> RelationWalker<'ast, 'db> {
+    pub(crate) fn refine(&self) -> RefinedRelationWalker<'ast, 'db> {
+        let relation = &self.db.relations.relations_storage[self.relation_id];
+
+        if relation.is_many_to_many() {
+            RefinedRelationWalker::ImplicitManyToMany(ImplicitManyToManyRelationWalker {
+                db: self.db,
+                relation_id: self.relation_id,
+            })
+        } else {
+            RefinedRelationWalker::Inline(InlineRelationWalker {
+                relation_id: self.relation_id,
+                db: self.db,
+            })
+        }
+    }
+}
+
+pub(crate) enum RefinedRelationWalker<'ast, 'db> {
+    Inline(InlineRelationWalker<'ast, 'db>),
+    ImplicitManyToMany(ImplicitManyToManyRelationWalker<'ast, 'db>),
+}
+
+#[derive(Copy, Clone)]
+pub(crate) struct InlineRelationWalker<'ast, 'db> {
+    relation_id: usize,
+    db: &'db ParserDatabase<'ast>,
+}
+
+impl<'ast, 'db> InlineRelationWalker<'ast, 'db> {
+    fn get(&self) -> &'db Relation<'ast> {
+        &self.db.relations.relations_storage[self.relation_id]
+    }
+
+    pub(crate) fn referencing_model(self) -> ModelWalker<'ast, 'db> {
+        self.db.walk_model(self.get().model_a)
+    }
+
+    pub(crate) fn referenced_model(self) -> ModelWalker<'ast, 'db> {
+        self.db.walk_model(self.get().model_b)
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn referenced_fields(self) -> Box<dyn Iterator<Item = ScalarFieldWalker<'ast, 'db>> + 'db> {
+        self.forward_relation_field()
+            .and_then(
+                |field: RelationFieldWalker<'ast, 'db>| -> Option<Box<dyn Iterator<Item = ScalarFieldWalker<'ast, 'db>>>> {
+                    field.referenced_fields().map(|fields| Box::new(fields) as Box<dyn Iterator<Item = ScalarFieldWalker<'ast, 'db>>>)
+                },
+            )
+            .unwrap_or_else(move || {
+                Box::new(
```

Comment:
This is wrong. This should only be taken from the unique criteria if we're in an m:n implicit relation.

Reply:
Nope. See the other comment, it does this only for m:n.